### DISCUSSION TITLE
Interessant

### DISCUSSION THREAD

**Discussion by fgamache1** [sle118/hayward_pool_heater#1] [General]
Salut. je cherche depuis quelques années à contrôler ma thermopompe par home assistant. J'avais vue le travail de njanik sur son controleur et j'ai essayer d'adapter au mien. Sans succès. :( Mais je suis depuis le début ton travail sur ce projet, et je pense qu'on pourrait s'aider mutuellement. 

Je vois que tu es également à Québec. Ca simplifie les trucs. Laisse moi savoir si tu es intéressé.

**Comment by sle118** [comment]
Tres certainement, oui. Je peaufine le code en ce moment pour permettre de mettre en evidence les changements entre chaque paquets d'un meme type. Ça devrait me permettre de mieux comprendre les différents "set points" par exemple. Le manuel technique en specifie 3, qui sont pour le refroidissement, le chauffage et le mode automatique. En plus de ca, il y a des paramètres régissant les ecarts minimums, etc. En manipulant les configs dans le menu user de l'appareil, je serai en mesure de localiser precisement ou chacun se loge.

**Comment by sle118** [comment]
I just committed the most recent code and the technical manual which explains how to get to the user settings in case you want to give it a try.

**Comment by sle118** [comment]
> Salut. je cherche depuis quelques années à contrôler ma thermopompe par home assistant. J'avais vue le travail de njanik sur son controleur et j'ai essayer d'adapter au mien. Sans succès. :( Mais je suis depuis le début ton travail sur ce projet, et je pense qu'on pourrait s'aider mutuellement.
> 
> Je vois que tu es également à Québec. Ca simplifie les trucs. Laisse moi savoir si tu es intéressé.

I'll continue in English here as this increase the chances that others would join. One of the things I could do with the project is to refactor it so protocol mappings and board specific mappings would be contained in a "codec" layer and a canonical layer each with the following goal:
- CODEC: decodes and validates payloads, encodes payloads for sending commands, maps data to and from the canonical model. This is also the level at which the "climate" capabilities would be reported to the climate component in ESPHome. For example, which modes are supported by the unit (HEAT,COOL,AUTO,etc).  
- Canonical: this is a generic data model with members for typical parameters of a pool heat pump like set points, inlet temp, out temp, coil temp, etc. Members are std::optional<> which allows the top integration layer (i.e. the ESPHome climate component) to know if a given value was observed on the bus by the codec. 

The climate.py would implement a list containing all supported models, from which users could choose, and the underlying code generator would instantiate the proper codec and pass it down to the bus driver.

I wouldn't go as far as allowing a different bus driver altogether, as I think the project should concentrate on the boards implementing similar protocols. Worst case is we could have the codec provide all timing info to the bus driver (e.g. what are the pulse length for a binary 1, etc).

This is something I could put on the roadmap for the upcoming cold winters.

Note: if anyone has a dead PCB, it would certainly help in many ways to analyze the protocols. This is something we could discuss in private

**Comment by sle118** [comment]
For anyone interested, the latest code that was pushed allows the creation of arbitrary frames using dependency injection, which ease the implementation for anyone willing to support other types of heat pumps. For now, though, the packet structure has to align with 1-10-1 (id, data, chksum) or 1-7-1 for short frames. If needed, though, it would be possible to get around that issue by moving the frame registry to its own class, opening the door for much more flexibility. I didn't need to go there, but it is doable.
